packages/task/src/browser/task-service.ts

```diff
@@ -727,34 +730,48 @@ export class TaskService implements TaskConfigurationClient {
+        console.debug('got lock');
 
-        // check if the task is active
-        const matchedRunningTaskInfo = runningTasksInfo.find(taskInfo => {
-            const taskConfig = taskInfo.config;
-            return this.taskDefinitionRegistry.compareTasks(taskConfig, task);
-        });
-        if (matchedRunningTaskInfo) { // the task is active
-            const taskName = this.taskNameResolver.resolve(task);
-            const terminalId = matchedRunningTaskInfo.terminalId;
-            if (terminalId) {
-                const terminal = this.terminalService.getByTerminalId(terminalId);
-                if (terminal) {
-                    if (TaskOutputPresentation.shouldSetFocusToTerminal(task)) { // assign focus to the terminal if presentation.focus is true
-                        this.terminalService.open(terminal, { mode: 'activate' });
-                    } else if (TaskOutputPresentation.shouldAlwaysRevealTerminal(task)) { // show the terminal but not assign focus
-                        this.terminalService.open(terminal, { mode: 'reveal' });
+        try {
+            const runningTasksInfo: TaskInfo[] = await this.getRunningTasks();
+            // check if the task is active
+            const matchedRunningTaskInfo = runningTasksInfo.find(taskInfo => {
+                const taskConfig = taskInfo.config;
+                return this.taskDefinitionRegistry.compareTasks(taskConfig, task);
+            });
+            console.debug(`running task ${JSON.stringify(task)}, already running = ${!!matchedRunningTaskInfo}`);
+
+            if (matchedRunningTaskInfo) { // the task is active
+                this.taskStartingLock.release(lockToken);
+                const taskName = this.taskNameResolver.resolve(task);
+                const terminalId = matchedRunningTaskInfo.terminalId;
+                if (terminalId) {
+                    const terminal = this.terminalService.getByTerminalId(terminalId);
+                    if (terminal) {
+                        if (TaskOutputPresentation.shouldSetFocusToTerminal(task)) { // assign focus to the terminal if presentation.focus is true
+                            this.terminalService.open(terminal, { mode: 'activate' });
+                        } else if (TaskOutputPresentation.shouldAlwaysRevealTerminal(task)) { // show the terminal but not assign focus
+                            this.terminalService.open(terminal, { mode: 'reveal' });
+                        }
                     }
                 }
+                const selectedAction = await this.messageService.info(`The task '${taskName}' is already active`, 'Terminate Task', 'Restart Task');
+                if (selectedAction === 'Terminate Task') {
+                    await this.terminateTask(matchedRunningTaskInfo);
+                } else if (selectedAction === 'Restart Task') {
+                    return this.restartTask(matchedRunningTaskInfo, option);
+                }
```

Comment:
Yes, that would be possible. The behavior has not changed in that respect. As I've mentioned in the description, some call sites consider it an error if no task is started, so it's safer to preserve that behavior. 
Because otherwise, the operation is stopped until the user clicks on the dialog, which might be much later. Starting a task at that point would be more surprising than stacked dialogs.